sys/kern/rman.c

```diff
@@ -1,83 +1,252 @@
+      if (rg->rg_start <= s->rg_end && rg->rg_end >= s->rg_start)
+        panic("overlapping regions");
+
+      /* Check for any overlap with the next region. */
+      t = TAILQ_NEXT(s, rg_link);
+      if (t) {
+        if (rg->rg_start <= t->rg_end && rg->rg_end >= t->rg_start)
+          panic("overlaping regions");
+
+        /*
+         * See if this region can be merged with the next region.
+         * If not, clear the pointer.
+         */
+        if (rg->rg_end + 1 != t->rg_start)
+          t = NULL;
+      }
+
+      /* See if we can merge with the current region. */
+      if (rg->rg_start == s->rg_end + 1) {
+        /* Can we merge all 3 regions? */
+        if (t) {
+          s->rg_end = t->rg_end;
+          TAILQ_CONCAT(&s->rg_resources, &t->rg_resources, r_link);
+          TAILQ_REMOVE(&rm->rm_regions, t, rg_link);
+          kfree(M_RES, t);
+          kfree(M_RES, rg);
+        } else {
+          s->rg_end = rg->rg_end;
+          kfree(M_RES, rg);
+        }
+      } else if (t) {
+        /* We can merge with just the next region. */
+        t->rg_start = rg->rg_start;
+        kfree(M_RES, rg);
+      } else if (s->rg_end < rg->rg_start) {
+        TAILQ_INSERT_AFTER(&rm->rm_regions, s, rg, rg_link);
+      } else {
+        TAILQ_INSERT_BEFORE(s, rg, rg_link);
+      }
+    }
+  }
+}
+
+void rman_init_from_resource(rman_t *rm, const char *name, resource_t *r) {
+  rman_init(rm, name);
+  rman_manage_region(rm, r->r_start, r->r_end);
+}
 
-  /* Adjust search boundaries if needed. */
-  start = max(start, rm->rm_start);
-  end = min(end, rm->rm_end);
+void rman_fini(rman_t *rm) {
+  WITH_MTX_LOCK (&rm->rm_lock) {
+    rman_region_t *rg;
+
+    while (!TAILQ_EMPTY(&rm->rm_regions)) {
+      rg = TAILQ_FIRST(&rm->rm_regions);
+      TAILQ_REMOVE(&rm->rm_regions, rg, rg_link);
+      if (!TAILQ_EMPTY(&rg->rg_resources))
+        panic("non-empty region");
+      kfree(M_RES, rg);
+    }
+  }
+  /* TODO: destroy the `rm_lock` after implementing `mtx_desotry`. */
+}
+
+static bool rman_region_reserve_resource(rman_region_t *rg, rman_addr_t start,
+                                         rman_addr_t end, size_t count,
+                                         resource_t *r) {
+  resource_t *s, *t;
+  rman_addr_t rstart, rend, amask;
+  res_flags_t flags = r->r_flags;
+
+  assert(mtx_owned(&rg->rg_rman->rm_lock));
 
-  /* Fits before the first resource on the list ? */
-  resource_t *first_r = TAILQ_FIRST(&rm->rm_resources);
-  rman_addr_t first_start = first_r ? first_r->r_start : rm->rm_end + 1;
-  if (align(start, bound) + count <= first_start) {
-    *start_p = align(start, bound);
+  amask = RMAN_ALIGNMENT_GET(flags);
+  start = RMAN_ALIGN(max(rg->rg_start, start), amask);
+  end = min(rg->rg_end, end);
+  if (end - count + 1 < start)
+    return false;
+
+  /* Does the resource fit before the first resource in the region? */
+  s = TAILQ_FIRST(&rg->rg_resources);
+  if (!s || (s && s->r_start - count >= start)) {
```

Comment:
Sometimes you are writing `!expr` sometimes `expr == NULL`. Please be consistent.